ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -250,58 +255,81 @@ void close() throws IOException {
+        numSegments);
+    File segmentFile = allocate(segmentNameAndFilename.getSecond(), segmentCapacity);
     // to be backwards compatible, headers are not written for a log segment if it is the only log segment.
-    LogSegment segment = new LogSegment(firstSegmentName, segmentFile, segmentCapacity, metrics, numSegments > 1);
-    segmentsByName.put(segment.getName(), segment);
+    return new LogSegment(segmentNameAndFilename.getFirst(), segmentFile, segmentCapacity, metrics, isLogSegmented);
   }
 
   /**
-   * Loads segment files and creates {@link LogSegment} instances.
+   * Creates {@link LogSegment} instances from {@code segmentFiles}.
    * @param segmentFiles the files that form the segments of the log.
-   * @param totalCapacity the total capacity of the log. This is used only if this is a single segment log.
+   * @return {@code List} of {@link LogSegment} instances corresponding to {@code segmentFiles}.
    * @throws IOException if there is an I/O error loading the segment files or creating {@link LogSegment} instances.
    */
-  private void loadSegments(File[] segmentFiles, long totalCapacity) throws IOException {
-    long totalSegments = -1;
+  private List<LogSegment> getSegmentsToLoad(File[] segmentFiles) throws IOException {
+    List<LogSegment> segments = new ArrayList<>(segmentFiles.length);
     for (File segmentFile : segmentFiles) {
       String name = LogSegmentNameHelper.nameFromFilename(segmentFile.getName());
       logger.info("Loading segment with name [{}]", name);
       LogSegment segment;
       if (name.isEmpty()) {
-        totalSegments = 1;
         // for backwards compatibility, a single segment log is loaded by providing capacity since the old logs have
         // no headers
-        segment = new LogSegment(name, segmentFile, totalCapacity, metrics, false);
+        segment = new LogSegment(name, segmentFile, capacityInBytes, metrics, false);
       } else {
         segment = new LogSegment(name, segmentFile, metrics);
-        totalSegments = totalSegments == -1 ? totalCapacity / segment.getCapacityInBytes() : totalSegments;
       }
       logger.info("Segment [{}] has capacity of {} bytes", name, segment.getCapacityInBytes());
+      segments.add(segment);
+    }
+    return segments;
+  }
+
+  /**
+   * Initializes the log.
+   * @param segmentsToLoad the {@link LogSegment} instances to include as a part of the log.
+   * @param segmentCapacityInBytes the capacity of a single {@link LogSegment}.
+   * @throws IOException if there is any I/O error during initialization.
+   */
+  private void initialize(List<LogSegment> segmentsToLoad, long segmentCapacityInBytes) throws IOException {
+    if (segmentsToLoad.size() == 0) {
+      // bootstrapping log.
+      segmentsToLoad = Collections.singletonList(checkArgsAndGetFirstSegment(segmentCapacityInBytes));
+    }
+
+    long totalSegments = -1;
+    for (LogSegment segment : segmentsToLoad) {
```

Comment:
can you set the value for totalSegments outside the for loop as one segment is good enough to find the value. Would prefer to have for loop to populate just segmentsByName